core/src/nodes/collection.rs

```diff
@@ -464,3 +464,382 @@ impl<'a, TInEvent> PeerMut<'a, TInEvent> {
+        let muxer = DummyMuxer::new();
+        let task_inner_fut = future::ok((peer_id.clone(), muxer));
+        let mut handler = Handler::default();
+        handler.next_states = vec![HandlerState::Ready(None)]; // triggered when sending a NextState event
+
+        cs.lock().add_reach_attempt(task_inner_fut, handler);
+        let mut rt = Builder::new().core_threads(1).build().unwrap();
+
+        // Kick it off
+        let cs_fut = cs.clone();
+        rt.block_on(future::poll_fn(move || -> Poll<_, ()> {
+            let mut cs = cs_fut.lock();
+            assert_matches!(cs.poll(), Async::NotReady);
+            // send an event so the Handler errors in two polls
+            cs.broadcast_event(&InEvent::NextState);
+            Ok(Async::Ready(()))
+        })).expect("tokio works");
+
+        // Accept the new node
+        let cs_fut = cs.clone();
+        rt.block_on(future::poll_fn(move || -> Poll<_, ()> {
+            let mut cs = cs_fut.lock();
+            // NodeReached, accept the connection so the task transitions from Pending to Connected
+            assert_matches!(cs.poll(), Async::Ready(CollectionEvent::NodeReached(reach_ev)) => {
+                reach_ev.accept();
+            });
+            Ok(Async::Ready(()))
+        })).expect("tokio works");
+
+        assert!(cs.lock().has_connection(&peer_id));
+
+        // Next poll, the Handler returns Async::Ready(None) because of the
+        // NextState message sent before.
+        let cs_fut = cs.clone();
+        rt.block_on(future::poll_fn(move || -> Poll<_, ()> {
+            let mut cs = cs_fut.lock();
+            // Node is closed normally: TaskClosed, Ok(())
+            assert_matches!(cs.poll(), Async::Ready(CollectionEvent::NodeClosed{ peer_id: peer_id_in_event }) => {
+                assert_eq!(peer_id_in_event, peer_id);
+            });
+            Ok(Async::Ready(()))
+        })).expect("tokio works");
+    }
+
+    #[test]
+    fn interrupting_a_pending_connection_attempt_is_ok() {
+        let mut cs = TestCollectionStream::new();
+        let fut = future::empty();
+        let reach_id = cs.add_reach_attempt(fut, Handler::default());
+        let interrupt = cs.interrupt(reach_id);
+        assert!(interrupt.is_ok());
+    }
+
+    #[test]
+    fn interrupting_a_pending_connection_with_invalid_reach_attempt_id_is_err() {
+        let mut cs = TestCollectionStream::new();
+        let fut = future::empty();
+        let valid_reach_id = cs.add_reach_attempt(fut, Handler::default());
+
+        // Obtain a TaskId we can use to build a ReachAttemptId that is not in
+        // the list of tasks in the CollectionStream we're testing. Note that
+        // TaskIds with the same number are all considered equal, so TaskId(0)
+        // from one HandledNodesTask is == to TaskId(0) from another
+        // HandledNodesTask.
+        let invalid_reach_id = {
```

Comment:
I don't think that's a nice way of doing it. You're relying on the fact that the task ID is linear, which is not necessarily true.
A better test, IMO, is to interrupt the same task twice and check that it is an error the second time.

Reply:
The original idea with this test was this: prove that I can't misuse a `TaskId` from reach attempts belonging to different `CollectionStream`s, to interrupt one using a `TaskId` from the other. That's actually allowed though (and not really a concern we care about I'd say).

Interrupting twice is a slightly different thing, but exercise the same code path so I'll add that. 

And remove this test.